feat(PromisePipe): add default pipe for promises

jeffbcross opened the pull request:
Spec implementation are heavily borrowed from the async pipe.

jeffbcross committed:
feat(PromisePipe): add default pipe for promises

mhevery commented on modules/angular2/src/change_detection/change_detection.js:
```diff
@@ -38,10 +39,16 @@ export var async = [
   "keyValDiff" : keyValDiff,
-  "async" : async
+  "async" : async,
+  "promise": promise
```
It should piggyback on `async` and not be a new type.

vsavkin commented:
I think you should merge |promise and |async. 

LGTM

jeffbcross commented:
@vsavkin mind taking another look after merging with async? Recommend ignoring whitespace: https://github.com/angular/angular/pull/1579/files?w=0

jeffbcross commented:
@vsavkin updated again with async pipe including promise

mhevery commented on modules/angular2/src/change_detection/change_detection.js:
```diff
@@ -38,10 +39,16 @@ export var async = [
   "keyValDiff" : keyValDiff,
-  "async" : async
+  "async" : async,
+  "async": promise
```
You can't have two keys with the same name. I think instead of creating new `PromisePipFactory` you should just augment the `AsyncPipeFactory` to understand promises.

jeffbcross replied:
@mhevery I'm confused. I refactored to implement as you describe, with async supporting observable OR promise, but @vsavkin told me he intended for me to just have promise as another async handler, relying on `supports()` of both implementations to determine which one can handle the input. @vsavkin did I misunderstand you again?

jeffbcross replied:
But yes, I see that regardless of the correct way, this is incorrect code, committed in haste, given duplicate keys.

mhevery commented:
@vsavkin I am removing !merge and @lgtm as I think the refactoring is a lot more complicated and needs a final review.

vsavkin commented on modules/angular2/src/change_detection/change_detection.js:
```diff
@@ -38,10 +39,15 @@ export var async = [
   "iterableDiff" : iterableDiff,
   "keyValDiff" : keyValDiff,
-  "async" : async
+  "async" : [async, promise]
```
I am pretty sure we do not support nested array.  I think it won't work.

You can, however, replace it with:

```
export var async = [
  new AsyncPipeFactory(),
  new PromisePipeFactory(),
  new NullPipeFactory()
];
```

vsavkin commented on modules/angular2/test/change_detection/pipes/promise_pipe_spec.js:
```diff
@@ -0,0 +1,99 @@
+
+        PromiseWrapper.setTimeout(() => {
+          pipe.transform(completer.promise);
+          expect(pipe.transform(completer.promise)).toBe(NO_CHANGE);
```
I changed it around to make `one|pipe` closer to `pipe(one)`. The way it works right now is:

* If you return the same instance as before, change detection won't update the binding.
* If you return a new instance, change detection will update the binding.

So I removed NO_CHANGE, and replaced it with WrappedValue. 

See `async_pipe_spec` and `async_pipe` for more info.

naomiblack commented:
@vsavkin this is still failing Travis... can you update with status?

vsavkin commented:
>> @vsavkin this is still failing Travis... can you update with status?

@jeffbcross what is the status of this PR?

jeffbcross commented:
@vsavkin @naomiblack I need to update with your suggestions, but I wasn't sure if the failures were flakes or if I was doing something to make Dart angry. Should be able to wrap this up Monday morning.

jeffbcross committed:
WIP: update promise pipe to core changes, and fix implementation in default pipes

jeffbcross committed:
WIP: fix traceur bug and rewrite tests for wrapped value

jeffbcross committed:
WIP: rename AsyncPipe to ObservablePipe

jeffbcross commented:
@vsavkin changes I made:

 * Rename `AsyncPipe` & `AsyncPipeFactory` to `ObservablePipe` & `ObservablePipeFactory`
 * Renamed promise_pipe file to .ts
 * Updated tests to use new Wrapper instead of NO_CHANGE
 * Fixed a couple of DartAnalyzer complaints

Mind taking a look?

jeffbcross committed:
WIP: remove unused variables to make dart analyzer happy

jeffbcross committed:
WIP: clang formatting

jeffbcross committed:
WIP: fix clang formatting

jeffbcross commented:
@naomiblack @vsavkin merged at 5691063ba0c8ebfea3cab43f567e0e5d04f4e0fa. Thanks for the help, Victor!
